app/src/main/java/com/zeapo/pwdstore/autofill/v2/ui/AutofillFilterView.kt

```diff
@@ -0,0 +1,171 @@
+                    this.typeImage.setImageResource(R.drawable.ic_action_secure)
+                }
+                onClick {
+                    setResponse(item)
+                    finish()
+                }
+            }
+        }
+
+        dataSource.set(getLastPasswordsList())
+
+        search.afterTextChanged {
+            dataSource.clear()
+            if (it.isEmpty()) {
+                dataSource.set(getLastPasswordsList())
+            } else {
+                recursiveFilter(it, null)
+            }
+        }
+
+        overlay.setOnClickListener {
+            setResponse(null)
+            finish()
+        }
+    }
+
+    private fun getLastPasswordsList(): List<PasswordItem> {
+        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N_MR1) {
+            val shortcutManager = getSystemService<ShortcutManager>()
+            if (shortcutManager != null) {
+                val shortcuts = shortcutManager.dynamicShortcuts
+                shortcuts.map {
+                    PasswordItem.newPassword(it.shortLabel.toString(), File(it.longLabel.toString()), File("/"))
+                }
+            } else {
+                emptyList()
+            }
+        } else {
+            emptyList()
+        }
+    }
+
+    private fun getStructure(): StructureParser.Result? {
+        val structure = intent.getParcelableExtra<AssistStructure>(AutofillManager.EXTRA_ASSIST_STRUCTURE)
+        return when {
+            structure != null -> StructureParser(structure).parse()
+            else -> null
+        }
+    }
+
+    private fun setResponse(item: PasswordItem?) {
+
+        if (item == null) {
+            setResult(RESULT_CANCELED)
+            return
+        }
+
+        val structure = getStructure() ?: return
+        val replyIntent = Intent()
+
+        // TODO Replace the Entry by the password decoded from the user choice
+        val response = FillResponse
+                .Builder()
+                .addDataset(AutofillUtils.buildDataset(this, item.longName, "Password", structure))
+                .build()
+
+        replyIntent.putExtra(AutofillManager.EXTRA_AUTHENTICATION_RESULT, response)
+
+        setResult(RESULT_OK, replyIntent)
+    }
+
+    @SuppressLint("DefaultLocale")
+    private fun recursiveFilter(filter: String, dir: File?) {
+        // on the root the pathStack is empty
+        val passwordItems = if (dir == null) {
+            PasswordRepository.getPasswords(PasswordRepository.getRepositoryDirectory(this)!!, getSortOrder())
+        } else {
+            PasswordRepository.getPasswords(dir, PasswordRepository.getRepositoryDirectory(this)!!, getSortOrder())
+        }
+
+        for (item in passwordItems) {
+            if (item.type == PasswordItem.TYPE_CATEGORY) {
+                recursiveFilter(filter, item.file)
+            }
+
+            val matches = ("${item.file.absolutePath}/$item").toLowerCase().contains(filter.toLowerCase())
```

Comment:
I guess we can easily implement fuzzy search instead of simple contains condition here.